backend/src/apiserver/server/list_request_util.go

```diff
@@ -157,3 +158,27 @@ func deserializePageToken(pageToken string) (*common.Token, error) {
 	}
 	return &token, nil
 }
+
+// parseAPIFilter attempts to decode a base-64 encoded JSON-stringified api
+// filter object. An empty string is considered valid input, and equivalent to
+// the nil filter, which trivially does nothing.
+func parseAPIFilter(encoded string) (*api.Filter, error) {
+	if encoded == "" {
+		return nil, nil
+	}
+
+	errorF := func(err error) (*api.Filter, error) {
+		return nil, util.NewInvalidInputError("failed to parse valid filter from %q: %v", encoded, err)
+	}
+
+	b, err := base64.URLEncoding.DecodeString(encoded)
+	if err != nil {
+		return errorF(err)
+	}
+
+	f := &api.Filter{}
+	if err := jsonpb.UnmarshalString(string(b), f); err != nil {
+		return errorF(err)
+	}
+	return f, nil
```

Comment:
can we add additional validation to make sure the filter format is correct?
e.g. value missing, op missing etc.

Reply:
The api filter will get parsed into a filter.Filter object that can actually be used to do filtering. That parsing will sanity check to ensure the filter is well formed and return an error otherwise.